### DISCUSSION TITLE
Discussion about API changes

### DISCUSSION THREAD

**Discussion by fredlcore** [fredlcore/BSB-LAN#258] [Ideas]
Here you can discuss and suggest changes to the (JSON) API for use with external modules in other home automation systems.

**Comment by hacki11** [comment]
Speaking for iobroker, the user should not need to "migrate" iobroker-adapter settings himself depending on which bsb_lan firmware is used. It should just work out of the box no matter if new firmware, old firmware or the user did an upgrade with an existing adapter. 
Introducing such "migration" steps scenarios depending on user based settings should be avoided at all.

Thats why i need to distinguish which api-version is currently available. This version should be incremented when breaking changes are introduced.
[Api Versioning Strategies](https://www.xmatters.com/blog/devops/blog-four-rest-api-versioning-strategies/)
I think for bsb_lan this will not fit because we are not supporting multiple versions. But perhaps a solution like [Supported Versions](https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/dome_versions.htm) would be feasable.

An endpoint or a field somewhere, where the current api-version is available. Of course the location of the field or endpoint should _never_ change, because it will eliminate its own goal.

With this i can implement a new bsb_lan client supporting the new endpoints and maintain just one adapter version without getting incompatible with user who don't want to update bsb_lan firmware. Furthermore i still can bugfix older versions easily. I just have to check which api-verison is used (not existing for v1, and the endpoint for > v1) and select the right client.

**Reply by fredlcore** [reply]
Thank you for your input - we already have /JI which dumps the configuration of BSB-LAN in a JSON structure. Would it be sufficient if we just add another field here which carries the API version?
As for versioning, I would use the same semantic versioning strategy that we use for BSB-LAN's main code as well: 
Major version change: breaking change
Minor version change: Adjustments needed (the question is where: For the user or the external module maintainer?)
Patch level change: No change needed

Should we do it like this or do you have other suggestions?

**Reply by hacki11** [reply]
I think we can omit the Patch level, because if nothing changes, no increment is needed. So we will stick with 0 all the time.
Yes, a field in /JI would be ok for me if /JI and the location of the field will never change in future.

Major: Breaking change / Breaking backwards compatibility (e.g. removing /JA)
Minor version change: Non breaking change (e.g. adding a new endpoint/feature/bugfix which can be used)

**Reply by fredlcore** [reply]
Ok, @dukess, can you add a field like this? We could use a definement such as API_VERSION to be displayed there.

**Reply by dukess** [reply]
In first sight we can use existing /JI "version" field as API version but here no problem to add new field.
But there is a problem here: /JI needs to be requested regularly to avoid out-of-sync version numbers from the server and client when updating the firmware or version field should be added to each JSON (/JQ, /JC, /JK...). Or this problem can be ignored?

**Reply by hacki11** [reply]
Good point, but if an endpoint is removed that version can not be read anymore. The version should not be part of the data, because it versions the data.

**Reply by hacki11** [reply]
It should be possible to request the version often and fast. If /JI contains too much data, there should be introduced an endpoint just for the version topic. (/JI is also data the version versions and should not be mixed, imho)

**Reply by dukess** [reply]
Clear. In this case adding new /JV command with payload {"version": "x.x" } looks better. Should we calculate API version separately or can use application version?

**Reply by hacki11** [reply]
Normally it is independent of application version. Totally fine to use separate version here

**Reply by dukess** [reply]
Done: https://github.com/fredlcore/bsb_lan/pull/260
Payload: {"api_version": "x.y" } 
Update: initial version is "2.0"

**Comment by hacki11** [comment]
Can you tell me the replacement of /JA? I'm fetching average values in json format here.

**Reply by fredlcore** [reply]
Parameters 20050-20099

**Reply by hacki11** [reply]
Does the JQ=20050,.. json response include the original ID?

**Reply by dukess** [reply]
Nope.
You can query /JI and get IDs from "averages" field.
or /JL from object with "id" == 13. Now /JL is more volatile but in future we can add text identifiers for parameters in additional to numbers for more readability.
